digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala

```diff
@@ -0,0 +1,444 @@
+    val (shortSegments, linearSegments) = lanes.partition(a => (a.endMeasure - a.startMeasure) < MinAllowedLength && roadLink.length >= MinAllowedLength )
+    val expiredLaneIds = shortSegments.map(_.id).toSet
+
+    (linearSegments, changeSet.copy(expiredLaneIds = changeSet.expiredLaneIds ++ expiredLaneIds.filterNot(_ == 0L)))
+  }
+
+  private def adjustAssets(roadLink: RoadLink, lanes: Seq[PersistedLane], changeSet: ChangeSet): (Seq[PersistedLane], ChangeSet) = {
+    lanes.foldLeft((Seq[PersistedLane](), changeSet)){
+      case ((resultAssets, change), lane) =>
+        val (asset, adjustmentsMValues) = adjustAsset(lane, roadLink)
+        (resultAssets ++ Seq(asset), change.copy(adjustedMValues = change.adjustedMValues ++ adjustmentsMValues))
+    }
+  }
+
+  private def adjustAsset(lane: PersistedLane, roadLink: RoadLink): (PersistedLane, Seq[MValueAdjustment]) = {
+    val roadLinkLength = GeometryUtils.geometryLength(roadLink.geometry)
+    val adjustedStartMeasure = if (lane.startMeasure < AllowedTolerance && lane.startMeasure > MaxAllowedError) Some(0.0) else None
+    val endMeasureDifference: Double = roadLinkLength - lane.endMeasure
+    val adjustedEndMeasure = if (endMeasureDifference < AllowedTolerance && endMeasureDifference > MaxAllowedError) Some(roadLinkLength) else None
+    val mValueAdjustments = (adjustedStartMeasure, adjustedEndMeasure) match {
+      case (None, None) => Nil
+      case (s, e)       => Seq(MValueAdjustment(lane.id, lane.linkId, s.getOrElse(lane.startMeasure), e.getOrElse(lane.endMeasure)))
+    }
+    val adjustedAsset = lane.copy(
+      startMeasure = adjustedStartMeasure.getOrElse(lane.startMeasure),
+      endMeasure = adjustedEndMeasure.getOrElse(lane.endMeasure))
+    (adjustedAsset, mValueAdjustments)
+  }
+
+
+  private def adjustLanesSideCodes(roadLink: RoadLink, lanes: Seq[PersistedLane], changeSet: ChangeSet): (Seq[PersistedLane], ChangeSet) = {
+
+    if(lanes.isEmpty )
+      return (lanes, changeSet)
+
+    val lanesToProcess = lanes.filter(_.linkId == roadLink.linkId)
+    val baseLane = lanesToProcess.sortBy(_.laneCode).head
+    val baseProps = baseLane.attributes.properties.filterNot(_.publicId == "lane_code")
+
+    roadLink.trafficDirection match {
+
+      case TrafficDirection.BothDirections =>
+
+        val mainLanes = ( lanesToProcess.exists(_.laneCode == 11),  lanesToProcess.exists(_.laneCode == 21) )
+
+        val toAdd = mainLanes match {
+          case (true, true) => Seq()
+          case (false, true) => Seq(
+                            PersistedLane(0L, roadLink.linkId, SideCode.TowardsDigitizing.value, 11, baseLane.municipalityCode,
+                          0, roadLink.length, None, None, None, None, expired = false, roadLink.vvhTimeStamp, None,
+                          LanePropertiesValues(baseProps ++ Seq(LaneProperty("lane_code", Seq(LanePropertyValue(11))))))
```

Comment:
Even though this is only a constructor the code would be cleaner if we had a local function to create PersistedLane and called that each time with different parameters

Reply:
Done